packages/eslint-plugin/src/rules/no-mixed-enums.ts

```diff
@@ -0,0 +1,222 @@
+          definition.node.range[0] < node.range[0] &&
+          definition.node.members.length > 0
+        ) {
+          found.previousSibling = definition.node;
+          break;
+        }
+      }
+
+      while (scope) {
+        scope.set.get(name)?.defs?.forEach(definition => {
+          if (definition.type === DefinitionType.ImportBinding) {
+            found.imports.push(definition.node);
+          }
+        });
+
+        scope = scope.upper;
+      }
+
+      return found;
+    }
+
+    function getAllowedTypeForNode(node: ts.Node): AllowedType | undefined {
+      return tsutils.isTypeFlagSet(
+        typeChecker.getTypeAtLocation(node),
+        ts.TypeFlags.StringLike,
+      )
+        ? ts.TypeFlags.String
+        : ts.TypeFlags.Number;
+    }
+
+    function getTypeFromImported(
+      imported: TSESTree.Node,
+    ): AllowedType | undefined {
+      const type = typeChecker.getTypeAtLocation(
+        parserServices.esTreeNodeToTSNodeMap.get(imported),
+      );
+
+      const valueDeclaration = type.getSymbol()?.valueDeclaration;
+      if (
+        !valueDeclaration ||
+        !ts.isEnumDeclaration(valueDeclaration) ||
+        valueDeclaration.members.length === 0
+      ) {
+        return undefined;
+      }
+
+      return getAllowedTypeForNode(valueDeclaration.members[0]);
+    }
+
+    function getMemberType(
+      member: TSESTree.TSEnumMember,
+    ): AllowedType | ts.TypeFlags.Unknown | undefined {
+      if (!member.initializer) {
+        return undefined;
+      }
+
+      switch (member.initializer.type) {
+        case AST_NODE_TYPES.Literal:
+          switch (typeof member.initializer.value) {
+            case 'number':
+              return ts.TypeFlags.Number;
+            case 'string':
+              return ts.TypeFlags.String;
+            default:
+              return ts.TypeFlags.Unknown;
+          }
+
+        case AST_NODE_TYPES.TemplateLiteral:
+          return ts.TypeFlags.String;
+
+        default:
+          return getAllowedTypeForNode(
+            parserServices.esTreeNodeToTSNodeMap.get(member.initializer),
+          );
+      }
+    }
+
+    function getDesiredTypeForDefinition(
+      node: TSESTree.TSEnumDeclaration,
+    ): AllowedType | ts.TypeFlags.Unknown | undefined {
+      const { imports, previousSibling } = collectNodeDefinitions(node);
+
+      // Case: Merged ambiently via module augmentation
+      // import { MyEnum } from 'other-module';
+      // declare module 'other-module' {
+      //   enum MyEnum { A }
+      // }
+      for (const imported of imports) {
+        const typeFromImported = getTypeFromImported(imported);
+        if (typeFromImported !== undefined) {
+          return typeFromImported;
+        }
+      }
+
+      // Case: Multiple enum declarations in the same file
+      // enum MyEnum { A }
+      // enum MyEnum { B }
+      if (previousSibling) {
+        return getMemberType(previousSibling.members[0]);
+      }
```

Comment:
we should use scope analysis here, similar to the fragment I shared below:

```ts
const enumDefinitions = context.getScope()?.upper?.set.get(enumName)?.defs;
if (enumDefinitions?.length === 1) {
  // there is exactly one definition so we're safe
  return orWhatever;
}
```

Reply:
Sorry @bradzacher I don't follow - are you suggesting I replace this `if` statement, or...?

The `scope.upper?.set.get(name)?.defs` is used inside `collectNodeDefinitions` for collecting `found.previousSibling.`

Reply:
oh shit sorry - i didn't do a full read of the change and just looked locally.
i saw "previous sibling" and thought it meant "one fetched from the AST".
all good then